### ISSUE TITLE
ANTLR Lexer Grammar Notation

### ISSUE THREAD

**Issue by RexJaeschke** [dotnet/csharpstandard#230]
## Comments

Line-oriented comments are supported, and we should use them, as appropriate. See below for examples. (Also, my proposed resolution to integrating the unsafe code grammar uses comments.)

## Ranges and Regular-Expression Sets
 
Currently, we have

```ANTLR
Hex_Digit
    : '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9'
    | 'A' | 'B' | 'C' | 'D' | 'E' | 'F' | 'a' | 'b' | 'c' | 'd' | 'e' | 'f'
	;
```

Using the `..` range notation, this can be rewritten compactly as

```ANTLR
Hex_Digit
    : '0'..'9' | 'A'..'F' | 'a'..'f'
    ;
```

And using regular-expression set notation, this can be rewritten even more compactly as

```ANTLR
Hex_Digit
    : [0-9A-Fa-f]
    ;
```

The regular-expression approach seems attractive, but can compactness interfere with readability? Which of the following do you prefer?

```ANTLR
Input_Character
    : ~('\u000D' | '\u000A' | '\u0085' | '\u2028' | '\u2029')  // anything but New_Line_Character
    ;
```

```ANTLR
Input_Character
    : ~[\u000D\u000A\u0085\u2028\u2029]  // anything but New_Line_Character
    ;
```

In at least one case, the regular-expression form *must* be used; for example:

```ANTLR
Whitespace
    : [\p{Zs}]	// Any character with Unicode class Zs
    ...
    ;
```

The notation `\p{...}` allows Unicode character properties to be specified. (Note this is *not* covered in the ANTLR book, as this feature came later.) It can *only* occur inside a set, even though it's a set of only 1 choice!

**Question: We currently do *not* use ranges or sets. Should we? If so, what are the guiding rules for choosing either of those over the existing approach?**

## Using Escape Sequences

A terminal is written as `'...'` where, typically, characters that are non-printable or difficult/impossible to enter on a keyboard are written as Unicode values. For example, `'A'` and `'\u0041'` both represent "Latin Capital letter A," while `'\t'` and `'\u0009'` both represent "Horizontal tab."

Currently, we have

```ANTLR
New_Line_Character
    : '<Carriage return character (U+000D)>'
    | '<Line feed character (U+000A)>'
    | '<Next line character (U+0085)>'
    | '<Line separator character (U+2028)>'
    | '<Paragraph separator character (U+2029)>'
    ;
```

As part of the resolution to Proposal 3 in Issue https://github.com/ECMA-TC49-TG2/csharpstandard/issues/37, we might chose to rewrite this, as follows:

```ANTLR
New_Line_Character
    : '\u000D'	// Carriage return
    | '\u000A'	// Line feed
    | '\u0085'	// Next line
    | '\u2028'	// Line separator
    | '\u2029'	// Paragraph separator
    ;
```

This approach makes the Unicode value clear, but puts it in the terminal rather than in the comment

Given that ANTLR recognizes the following escape sequences, `\n`, `\r`, `\b`, `\t`, and `\f`, this rule could be written instead as

```ANTLR
New_Line_Character
    : '\r'	// Carriage return
    | '\n'	// Line feed
    | '\u0085'	// Next line
    | '\u2028'	// Line separator
    | '\u2029'	// Paragraph separator
    ;
```

**Question: How about this for a guiding principle? First choice is to write any printable Latin character literally; second is to use an ANTLR escape; Thirdly, use a Unicode sequence.**

We currently have the following:

```ANTLR
Underscore_Character
    : '<_ the underscore character (U+005F)>'
    | ...
    ; 
```

As part of the resolution to Proposal 3 in Issue https://github.com/ECMA-TC49-TG2/csharpstandard/issues/37, we might chose to rewrite this, as follows:

```ANTLR
Underscore_Character
    : '_'    // Underscore character (U+005F)
    | ...
    ; 
```

keeping the Unicode value in the comment, as there likely are more than one Unicode glyph that look like an underscore. That said, we do *not* give explicit Unicode values for terminals like `'/'`, `'*'`, `'+'`, and `'-'`, which might also have look-alike glyphs.

**Question: If a character is not written as a Unicode sequence, when should we state a character's Unicode value in a comment?**

**Comment by RexJaeschke** [comment]
Above, @jskeet, @Nigel-Ecma, and I all suggested that we add a comment to show the Unicode value of non-alphanumeric terminal characters; for example:

```ANTLR
Underscore_Character
    : '_'    // underscore character (U+005F)
    | ...
    ; 
```

However, I now find in 7.2 Grammars|7.2.1 General, the following normative text:

> All terminal characters are to be understood as the appropriate Unicode character from the range U+0020 to U+007F, as opposed to any similar-looking characters from other Unicode character ranges.

This clearly covers *all* the cases we care about. Now while a comment on numerous lexer rule alternatives *would* make this overt for *every* case, some of the alternatives fill the whole line, in which case, any trailing comment would have to wrap to the next line. For example:

```ANTLR
Unicode_Escape_Sequence
    : '\\u' Hex_Digit Hex_Digit Hex_Digit Hex_Digit  // reverse solidus (U+005C)
    | '\\U' Hex_Digit Hex_Digit Hex_Digit Hex_Digit Hex_Digit Hex_Digit Hex_Digit Hex_Digit  // reverse solidus (U+005C)
    ;
```

We could shorten the comment by omitting the character's name, but that makes the following problematic:

```ANTLR
    : '@"' Verbatim_String_Literal_Character* '"'        // commercial at (U+0040), quotation mark (U+0022)
    ;
```

The comment indicates two different characters, but there are actually three with one occurring twice. If we remove the character names, what is left is incomplete and potentially confusing.

Who supports relying on the general statement in 7.2.1 and *not* having any comments on lexer rule alternatives to reinforce that?

**Comment by RexJaeschke** [comment]
Resolved by PR #261.
